Add redirection path in the error message of assert_response if response is :redirect

prathamesh-sonpatki opened the pull request:
- If the assert_response is checking for any non-redirect response like
  :success and actual response is :redirect then, the error message displayed was -
  
  ```
   Expected response to be a <success>, but was <302>
  ```
- This commit adds the redirect path to the error message of
  assert_response if the response is :redirect. So above message is changed to -
  
  ```
  Expected response to be a <success>, but was a redirect to <http://test.host/posts/lol>
  ```

rafaelfranca commented on actionpack/lib/action_dispatch/testing/assertions/response.rb:
```diff
@@ -77,6 +77,20 @@ def normalize_argument_to_redirection(fragment)
+
+          message
+        end
+
```
:scissors: this blank line

rafaelfranca commented on actionpack/lib/action_dispatch/testing/assertions/response.rb:
```diff
@@ -77,6 +77,20 @@ def normalize_argument_to_redirection(fragment)
+        def generate_response_message(type)
+          message = "Expected response to be a <#{type}>, but was"
+
+          if @response.response_code.to_s =~ /3[0-9][0-9]/
```
Maybe we have a method in response that tell us if it was a redirect?

prathamesh-sonpatki replied:
@rafaelfranca Correct. https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/testing/test_response.rb#L20 which aliases methods from Rack::Response::Helpers - https://github.com/rack/rack/blob/master/lib/rack/response.rb#L137

prathamesh-sonpatki commented:
@rafaelfranca Please take a look.

We get an error message like this now ->
```
Expected response to be a <200>, but was a redirect to <http://test.host/posts/lol>.
```

chrisccerami commented:
:+1: I would really appreciate this improved messaging. Usually after a failing test like this I need to then spend some extra time figuring out exactly where I'm redirecting too. The more informative messages can be the better.

prathamesh-sonpatki commented:
@rafaelfranca Any thoughts?

arthurnn commented:
I like this message better. I think we need to add a regression test for this tho?

prathamesh-sonpatki commented:
@arthurnn Added test here - https://github.com/prathamesh-sonpatki/rails/commit/bfe6bd873e999bd86e187aa3e63c8be117dcd80e. 

```
$ ruby -Ilib:test test/assertions/response_assertions_test.rb
Run options: --seed 43879

# Running:

....F

Finished in 0.144490s, 34.6045 runs/s, 179.9432 assertions/s.

  1) Failure:
ActionDispatch::Assertions::ResponseAssertionsTest#test_message_when_response_is_redirect_but_asserted_for_status_other_than_redirect [test/assertions/response_assertions_test.rb:74]:
--- expected
+++ actual
@@ -1 +1,2 @@
-"Expected response to be a <success>, but was a redirect to <http://test.host/posts/redirect/1>."
+"Expected response to be a <success>, but was a redirect to <http://test.host/posts/redirect/1>.
+Expected #<struct ActionDispatch::Assertions::ResponseAssertionsTest::FakeResponse response_code=:redirection, location=\"http://test.host/posts/redirect/1\"> to be successful?."


5 runs, 26 assertions, 1 failures, 0 errors, 0 skips
```

It's currently failing because call to [assert_predicate](https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/testing/assertions/response.rb#L32) is adding it's own error message to our error message.

Also noticed that we are comparing symbols with numbers here - https://github.com/rails/rails/blob/master/actionpack/test/assertions/response_assertions_test.rb#L10-L13 so the generated methods will always result in false. 

Current tests are passing error code as numbers - https://github.com/rails/rails/blob/master/actionpack/test/assertions/response_assertions_test.rb#L29, https://github.com/rails/rails/blob/master/actionpack/test/assertions/response_assertions_test.rb#L42

arthurnn commented:
> It's currently failing because call to assert_predicate is adding it's own error message to our error message.

that uses the message given, so if we are passing our message to the assert_predicate this should work just fine right?

prathamesh-sonpatki commented:
@arthurnn I also thought the same. But it prints our message and then prints it's own message.

For eg.

``` ruby
assert_predicate 1, :even?, "LOL"
LOL.
Expected 1 to be even?.
```

I tested this by adding above test in same file.

prathamesh-sonpatki commented:
Here is a test script https://gist.github.com/prathamesh-sonpatki/d929133a59586a15d04a

prathamesh-sonpatki commented:
Looks like the behavior of `assert_predicate` is expected - https://github.com/seattlerb/minitest/blob/c674fa648550d179c3a1440bc3790350f4892b2c/lib/minitest/assertions.rb#L283 and https://github.com/seattlerb/minitest/blob/c674fa648550d179c3a1440bc3790350f4892b2c/lib/minitest/assertions.rb#L483

prathamesh-sonpatki commented:
@arthurnn I fixed the test using `assert_match`.

chrisccerami commented:
:+1:

arthurnn commented:
looks good to me, can you rebase this against latest master(as there is conflicts) and i will merge it in.

thanks

prathamesh-sonpatki committed:
Add redirection path in the error message of assert_response if response is :redirect

- If the assert_response is checking for any non-redirect response like
  :success and actual response is :redirect then, the error message displayed
  was -

     Expected response to be a <success>, but was <302>

- This commit adds the redirect path to the error message of
  assert_response if the response is :redirect.
  So above message is changed to -

     Expected response to be a <success>, but was a redirect to <http://test.host/posts/lol>

prathamesh-sonpatki commented:
@arthurnn I rebased but there was no conflict. Also build failures seem to be unrelated to this change.

arthurnn commented:
thanks

matthewd commented on actionpack/lib/action_dispatch/testing/assertions/response.rb:
```diff
@@ -82,6 +84,19 @@ def normalize_argument_to_redirection(fragment)
+
+          if @response.redirection?
+            redirect_is = normalize_argument_to_redirection(@response.location)
+            message << " a redirect to <#{redirect_is}>"
```
Obscuring the type of redirect that occurred seems unhelpful... if we `assert_response 301`, then we could be told "Expected response to be a <301>, but was a redirect to <..>", which sounds a lot like "error: no error" without an indication that the response was a 302.

jonatack replied:
I agree. Submitted #22511 to show the 302.
